pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DictionaryBasedGroupKeyGenerator.java

```diff
@@ -831,6 +777,156 @@ private String getGroupKey(IntArray rawKey) {
     return groupKeyBuilder.toString();
   }
 
+  /**
+   * Fast int-to-int hashmap with {@link #INVALID_ID} as the default return value.
+   * <p>Different from {@link it.unimi.dsi.fastutil.ints.Int2IntOpenHashMap}, this map uses one single array to store
+   * keys and values to reduce the cache miss.
+   */
+  @VisibleForTesting
+  public static class IntGroupIdMap {
+    private static final float LOAD_FACTOR = 0.75f;
+
+    private int[] _keyValueHolder;
+    private int _capacity;
+    private int _mask;
+    private int _maxNumEntries;
+    private int _size;
+
+    public IntGroupIdMap() {
+      _capacity = 1 << 10;
+      int holderSize = _capacity << 1;
+      _keyValueHolder = new int[holderSize];
+      _mask = holderSize - 1;
+      _maxNumEntries = (int) (_capacity * LOAD_FACTOR);
+    }
+
+    public int size() {
+      return _size;
+    }
+
+    /**
+     * Returns the group id for the given raw key. Create a new group id if the raw key does not exist and the group id
+     * upper bound is not reached.
+     */
+    public int getGroupId(int rawKey, int groupIdUpperBound) {
+      // NOTE: Key 0 is reserved as the null key. Use (rawKey + 1) as the internal key because rawKey can never be -1.
+      int internalKey = rawKey + 1;
+      int index = (HashCommon.mix(internalKey) << 1) & _mask;
+      int key = _keyValueHolder[index];
+
+      // Handle hash hit separately for better performance
+      if (key == internalKey) {
+        return _keyValueHolder[index + 1];
+      }
+      if (key == 0) {
+        return _size < groupIdUpperBound ? addNewGroup(internalKey, index) : INVALID_ID;
```

Comment:
So `key == 0` implies we hashed to the empty array slot right ?

Reply:
Yes, 0 is reserved as the null key (check the comment on line 812)